File: .github/workflows/check-questions-sorted.py
#!/usr/bin/env python3
"""
Script to verify that questions in README files are sorted alphabetically by question text.

This script checks that question links in README.md and README.ru.md are sorted
in alphabetical order by their question text in the respective language, as specified in CONTRIBUTING.md.
"""

import re
import sys
from pathlib import Path
from typing import List, Tuple


def extract_question_links(readme_content: str) -> List[Tuple[str, str, str]]:
    """
    Extract question links from README content.

    Returns a list of tuples: (question_text, directory_name, full_line)
    """
    questions = []
    # Pattern to match markdown links to questions
    # Matches: - [Question text](./questions/directory-name/index.md)
    pattern = r'^-\s+\[(.+?)\]\(\./questions/([^/]+)/index(?:\.ru)?\.md\)\s*$'

    for line in readme_content.split('\n'):
        match = re.match(pattern, line.strip())
        if match:
            question_text = match.group(1)
            directory_name = match.group(2)
            questions.append((question_text, directory_name, line))

    return questions


def check_alphabetical_order(questions: List[Tuple[str, str, str]], filename: str) -> bool:
    """
    Check if questions are in alphabetical order by question text.

    Returns True if sorted correctly, False otherwise.
    """
    if not questions:
        print(f"Warning: No questions found in {filename}")
        return True

    question_texts = [q[0] for q in questions]
    sorted_texts = sorted(question_texts, key=str.lower)

    if question_texts == sorted_texts:
        print(f"✓ {filename}: Questions are sorted alphabetically ({len(questions)} questions)")
        return True
    else:
        print(f"✗ {filename}: Questions are NOT sorted alphabetically by question text")
        print("\nExpected order:")
        for text in sorted_texts:
            print(f"  - {text[:80]}{'...' if len(text) > 80 else ''}")
        print("\nActual order:")
        for text in question_texts:
            print(f"  - {text[:80]}{'...' if len(text) > 80 else ''}")

        # Find the first mismatch
        for i, (actual, expected) in enumerate(zip(question_texts, sorted_texts)):
            if actual != expected:
                print(f"\nFirst mismatch at position {i + 1}:")
                print(f"  Expected: {expected[:80]}{'...' if len(expected) > 80 else ''}")
                print(f"  Found: {actual[:80]}{'...' if len(actual) > 80 else ''}")
                break

        return False


def main():
    """Main function to check both README files."""
    script_dir = Path(__file__).parent.parent.parent
    readme_files = ['README.md', 'README.ru.md']

    all_sorted = True

    for readme_file in readme_files:
        readme_path = script_dir / readme_file

        if not readme_path.exists():
            print(f"Error: {readme_file} not found at {readme_path}")
            all_sorted = False
            continue

        content = readme_path.read_text(encoding='utf-8')
        questions = extract_question_links(content)

        is_sorted = check_alphabetical_order(questions, readme_file)
        all_sorted = all_sorted and is_sorted

    if all_sorted:
        print("\n✓ All README files have questions sorted alphabetically by question text")
        sys.exit(0)
    else:
        print("\n✗ Some README files have unsorted questions")
        print("\nPlease sort the questions alphabetically by question text in the respective language.")
        print("See CONTRIBUTING.md for details.")
        sys.exit(1)


if __name__ == '__main__':
    main()

File: .github/workflows/check-questions-sorted.mjs
#!/usr/bin/env node
/**
 * Script to verify that questions in README files are sorted alphabetically by question text.
 *
 * This script checks that question links in README.md and README.ru.md are sorted
 * in alphabetical order by their question text in the respective language, as specified in CONTRIBUTING.md.
 */

import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/**
 * Extract question links from README content.
 *
 * @param {string} readmeContent - The content of the README file
 * @returns {Array<{questionText: string, directoryName: string, fullLine: string}>} Array of question objects
 */
function extractQuestionLinks(readmeContent) {
  const questions = [];
  // Pattern to match markdown links to questions
  // Matches: - [Question text](./questions/directory-name/index.md)
  const pattern = /^-\s+\[(.+?)\]\(\.\/questions\/([^/]+)\/index(?:\.ru)?\.md\)\s*$/;

  for (const line of readmeContent.split('\n')) {
    const match = line.trim().match(pattern);
    if (match) {
      questions.push({
        questionText: match[1],
        directoryName: match[2],
        fullLine: line
      });
    }
  }

  return questions;
}

/**
 * Check if questions are in alphabetical order by question text.
 *
 * @param {Array<{questionText: string, directoryName: string, fullLine: string}>} questions - Array of question objects
 * @param {string} filename - Name of the README file being checked
 * @returns {boolean} True if sorted correctly, false otherwise
 */
function checkAlphabeticalOrder(questions, filename) {
  if (questions.length === 0) {
    console.log(`Warning: No questions found in ${filename}`);
    return true;
  }

  const questionTexts = questions.map(q => q.questionText);
  const sortedTexts = [...questionTexts].sort((a, b) => a.toLowerCase().localeCompare(b.toLowerCase()));

  if (JSON.stringify(questionTexts) === JSON.stringify(sortedTexts)) {
    console.log(`✓ ${filename}: Questions are sorted alphabetically (${questions.length} questions)`);
    return true;
  } else {
    console.log(`✗ ${filename}: Questions are NOT sorted alphabetically by question text`);
    console.log("\nExpected order:");
    for (const text of sortedTexts) {
      const truncated = text.length > 80 ? text.substring(0, 80) + '...' : text;
      console.log(`  - ${truncated}`);
    }
    console.log("\nActual order:");
    for (const text of questionTexts) {
      const truncated = text.length > 80 ? text.substring(0, 80) + '...' : text;
      console.log(`  - ${truncated}`);
    }

    // Find the first mismatch
    for (let i = 0; i < Math.min(questionTexts.length, sortedTexts.length); i++) {
      if (questionTexts[i] !== sortedTexts[i]) {
        console.log(`\nFirst mismatch at position ${i + 1}:`);
        const expectedTrunc = sortedTexts[i].length > 80 ? sortedTexts[i].substring(0, 80) + '...' : sortedTexts[i];
        const actualTrunc = questionTexts[i].length > 80 ? questionTexts[i].substring(0, 80) + '...' : questionTexts[i];
        console.log(`  Expected: ${expectedTrunc}`);
        console.log(`  Found: ${actualTrunc}`);
        break;
      }
    }

    return false;
  }
}

/**
 * Main function to check both README files.
 */
function main() {
  const scriptDir = join(__dirname, '..', '..');
  const readmeFiles = ['README.md', 'README.ru.md'];

  let allSorted = true;

  for (const readmeFile of readmeFiles) {
    const readmePath = join(scriptDir, readmeFile);

    let content;
    try {
      content = readFileSync(readmePath, 'utf-8');
    } catch (error) {
      console.log(`Error: ${readmeFile} not found at ${readmePath}`);
      allSorted = false;
      continue;
    }

    const questions = extractQuestionLinks(content);
    const isSorted = checkAlphabeticalOrder(questions, readmeFile);
    allSorted = allSorted && isSorted;
  }

  if (allSorted) {
    console.log("\n✓ All README files have questions sorted alphabetically by question text");
    process.exit(0);
  } else {
    console.log("\n✗ Some README files have unsorted questions");
    console.log("\nPlease sort the questions alphabetically by question text in the respective language.");
    console.log("See CONTRIBUTING.md for details.");
    process.exit(1);
  }
}

main();
